src/osd/osd_types.h

```diff
@@ -2745,32 +2792,60 @@ inline ostream& operator<<(ostream& out, const pg_log_t& log)
  *  kept in memory, as a supplement to pg_log_t
  *  also used to pass missing info in messages.
  */
+struct pg_missing_item_old {
+  eversion_t need, have;
+  pg_missing_item_old() {}
+  explicit pg_missing_item_old(eversion_t n) : need(n) {}  // have no old version
+  pg_missing_item_old(eversion_t n, eversion_t h) : need(n), have(h) {}
+
+  void encode(bufferlist& bl) const {
+    ::encode(need, bl);
+    ::encode(have, bl);
+  }
+  void decode(bufferlist::iterator& bl) {
+    ::decode(need, bl);
+    ::decode(have, bl);
+  }
+};
+WRITE_CLASS_ENCODER(pg_missing_item_old);
+
 struct pg_missing_item {
   eversion_t need, have;
+  ObjectCleanRegions clean_regions;
   pg_missing_item() {}
   explicit pg_missing_item(eversion_t n) : need(n) {}  // have no old version
-  pg_missing_item(eversion_t n, eversion_t h) : need(n), have(h) {}
+  pg_missing_item(eversion_t n, eversion_t h, bool old_style = false) : need(n), have(h) {
+    if (old_style) {
+      clean_regions.mark_data_region_dirty(0, (uint64_t)-1);
+      clean_regions.mark_omap_dirty();
+    }
+  }
 
   void encode(bufferlist& bl) const {
     ::encode(need, bl);
     ::encode(have, bl);
+    ::encode(clean_regions, bl);
   }
   void decode(bufferlist::iterator& bl) {
     ::decode(need, bl);
     ::decode(have, bl);
+    ::decode(clean_regions, bl);
```

Comment:
I don't understand why we can't transition this to use ENCODE_START etc.  You already must have switched the on-wire encode/decode code to use the features vector to do a conditional encode/decode (or it would crash old osds), and you must already be handling reading the old encoding from disk...